File: trivialkart/trivialkart-unity/Assets/Scripts/InputSDK/InputSDKRemappingListener.cs
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#if PLAY_GAMES_PC

using System.Text;
using Java.Lang;
using Java.Util;
using Google.Android.Libraries.Play.Games.Inputmapping;
using Google.Android.Libraries.Play.Games.Inputmapping.Datamodel;
using UnityEngine;
using UnityEngine.UI;

public class InputSDKRemappingListener : InputRemappingListenerCallbackHelper
{
    public override void OnInputMapChanged(InputMap inputMap)
    {
        Debug.Log("Received update on remapped controls.");
        if (inputMap.InputRemappingOption() == InputEnums.REMAP_OPTION_DISABLED)
        {
            return;
        }
        List<InputGroup> inputGroups = inputMap.InputGroups();
        for (int i = 0; i < inputGroups.Size(); i ++)
        {
            InputGroup inputGroup = inputGroups.Get(i);
            if (inputGroup.InputRemappingOption()
                    == InputEnums.REMAP_OPTION_DISABLED)
            {
                continue;
            }
            List<InputAction> inputActions = inputGroup.InputActions();
            for (int j = 0; j < inputActions.Size(); j ++)
            {
                InputAction inputAction = inputActions.Get(j);
                if (inputAction.InputRemappingOption()
                        != InputEnums.REMAP_OPTION_DISABLED)
                {
                    // Found action remapped by user
                    ProcessRemappedAction(inputAction);
                }
            }
        }
    }

    private void ProcessRemappedAction(InputAction remappedInputAction)
    {
        InputControls remappedInputControls =
                remappedInputAction.RemappedInputControls();
        List<Integer> remappedKeycodes = remappedInputControls.Keycodes();
        List<Integer> mouseActions = remappedInputControls.MouseActions();
        string version = remappedInputAction.InputActionId().VersionString();
        long remappedActionId = remappedInputAction.InputActionId().UniqueId();
        InputAction currentInputAction;
        if (string.IsNullOrEmpty(version)
                || string.Equals(
                version, InputSDKMappingProvider.INPUT_MAP_VERSION))
        {
            currentInputAction = GetCurrentVersionInputAction(remappedActionId);
        }
        else
        {
            Debug.Log("Detected version of used-saved input action " +
                "is different from current version");
            currentInputAction =
                GetCurrentVersionInputActionFromPreviousVersion(
                    remappedActionId, version);
        }
        if (currentInputAction == null)
        {
            Debug.LogError(string.Format(
                "Input Action with id {0} and version {1} not found",
                remappedActionId,
                string.IsNullOrEmpty(version) ? "UNKNOWN" : version));
            return;
        }
        InputControls originalControls = currentInputAction.InputControls();
        List<Integer> originalKeycodes = originalControls.Keycodes();

        Debug.Log(string.Format(
            "Found Input Action with id {0} remapped from key {1} to key {2}",
            remappedActionId,
            KeyCodesToString(originalKeycodes),
            KeyCodesToString(remappedKeycodes)));

        // Update HUD according to the controls of the user
        switch(remappedActionId)
        {
            case (int)InputSDKMappingProvider.InputEventIds.OPEN_GARAGE:
                GameObject.Find("GarageButton").GetComponentInChildren<Text>()
                    .text = string.Format(
                        "Garage({0})", GetLabelFromKeycode(remappedKeycodes));
                break;
            case (int)InputSDKMappingProvider.InputEventIds.OPEN_PGS:
                GameObject.Find("PGSButton").GetComponentInChildren<Text>()
                    .text = string.Format(
                        "PGS({0})", GetLabelFromKeycode(remappedKeycodes));
                break;
            case (int)InputSDKMappingProvider.InputEventIds.OPEN_STORE:
                GameObject.Find("StoreButton").GetComponentInChildren<Text>()
                    .text = string.Format(
                        "Store({0})", GetLabelFromKeycode(remappedKeycodes));
                break;
            default:
                Debug.Log("Skipping UI update because a non-UI control was remapped");
                break;
        }
    }

    private InputAction
            GetCurrentVersionInputAction(long inputActionId)
    {
        List<InputGroup> inputGroups =
            InputSDKMappingProvider.inputMap.InputGroups();
        for (int i = 0; i < inputGroups.Size(); i ++)
        {
            InputGroup inputGroup = inputGroups.Get(i);
            List<InputAction> inputActions = inputGroup.InputActions();
            for (int j = 0; j < inputActions.Size(); j ++)
            {
                InputAction inputAction = inputActions.Get(j);
                if (inputAction.InputActionId().UniqueId() == inputActionId)
                {
                    return inputAction;
                }
            }
        }
        return null;
    }

    private InputAction
            GetCurrentVersionInputActionFromPreviousVersion(
            long inputActionId, string version)
    {
        // TODO: add logic to this method considering the diff between your
        // current and previous InputMap.
        return null;
    }

    private string KeyCodesToString(List<Integer> keycodes)
    {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < keycodes.Size(); i ++)
        {
            Integer keycode = keycodes.Get(i);
            if (builder.Length > 0)
            {
                builder.Append(" + ");
            }
            builder.Append(keycode.IntValue());
        }
        return string.Format("({0})", builder.ToString());
    }

    private string GetLabelFromKeycode(List<Integer> keycodes)
    {
        // Access the keycode names through reflection
        var keycodeLabels = typeof(AndroidKeyCode).GetMembers();
        // Skip the first 4 reflected members: Equals, GetHashCode, GetType and
        // ToString
        int reservedMembers = 4;
        if (keycodes.Size() < 0 || keycodes.Size() > 1)
        {
            // Empty keycodes or multi-key keycodes are invalid remapped keys
            return "?";
        }
        int index = keycodes.Get(0).IntValue() + reservedMembers;
        if (index < reservedMembers || index >= keycodeLabels.Length)
        {
            // Unknown remapped key
            return "?";
        }
        string fullLabel = keycodeLabels[index].Name;
        if (!fullLabel.StartsWith("KEYCODE_"))
        {
            // Invalid operation
            return "?";
        }
        // Retrieve the label after the KEYCODE_ prefix of length 8
        string label = fullLabel.Substring(8);
        if (string.IsNullOrEmpty(label))
        {
            // Error during trimming operation
            return "?";
        }
        // Trim label to 1 char to prevent overlapping UI (example: SPACE -> S)
        return label.Substring(0, 1);
    }
}
#endif

File: agdk/agdktunnel/app/src/main/java/com/google/sample/agdktunnel/InputSDKRemappingListener.java
package com.google.sample.agdktunnel;

import android.util.Log;

import com.google.android.libraries.play.games.inputmapping.datamodel.InputGroup;
import com.google.android.libraries.play.games.inputmapping.datamodel.InputAction;
import com.google.android.libraries.play.games.inputmapping.datamodel.InputControls;
import com.google.android.libraries.play.games.inputmapping.datamodel.InputMap;
import com.google.android.libraries.play.games.inputmapping.datamodel.InputEnums;
import com.google.android.libraries.play.games.inputmapping.InputRemappingListener;
import java.util.List;
import java.util.Optional;

public class InputSDKRemappingListener implements InputRemappingListener {

    private static final String TAG = "InputSDKRemappingListener";

    @Override
    public void onInputMapChanged(InputMap inputMap) {
        Log.i(TAG, "Received update on input map changed.");
        if (inputMap.inputRemappingOption() == InputEnums.REMAP_OPTION_DISABLED) {
            return;
        }
        for (InputGroup inputGroup : inputMap.inputGroups()) {
            if (inputGroup.inputRemappingOption() == InputEnums.REMAP_OPTION_DISABLED) {
                continue;
            }
            for (InputAction inputAction : inputGroup.inputActions()) {
                if (inputAction.inputRemappingOption() != InputEnums.REMAP_OPTION_DISABLED &&
                        !inputAction.inputControls().equals(inputAction.remappedInputControls())) {
                    // Found InputAction remapped by user
                    processRemappedAction(inputAction);
                }
            }
        }
    }

    private void processRemappedAction(InputAction remappedInputAction) {
        // Get remapped action info
        InputControls remappedControls = remappedInputAction.remappedInputControls();
        List<Integer> remappedKeyCodes = remappedControls.keycodes();
        List<Integer> mouseActions = remappedControls.mouseActions();
        String version = remappedInputAction.inputActionId().versionString();
        long remappedActionId = remappedInputAction.inputActionId().uniqueId();
        Optional<InputAction> currentInputAction;
        if (version == null || version.isEmpty()
                || version.equals(InputSDKProvider.INPUTMAP_VERSION)) {
            currentInputAction = getCurrentVersionInputAction(remappedActionId);
        } else {
            Log.i(TAG,
                    "Detected version of user-saved input action defers from current version");
            currentInputAction = getCurrentVersionInputActionFromPreviousVersion(
                    remappedActionId, version);
        }
        if (!currentInputAction.isPresent()) {
            Log.e(TAG, String.format(
                    "can't find remapped input action with id %d and version %s",
                    remappedActionId, version == null || version.isEmpty() ? "UNKNOWN" : version));
            return;
        }
        InputControls originalControls = currentInputAction.get().inputControls();
        List<Integer> originalKeyCodes = originalControls.keycodes();

        Log.i(TAG, String.format(
                "Found input action with id %d remapped from key %s to key %s",
                remappedActionId,
                keyCodesToString(originalKeyCodes),
                keyCodesToString(remappedKeyCodes)));

        // TODO: game developer can make display changes to match controls used by the user.
    }

    private Optional<InputAction> getCurrentVersionInputAction(long inputActionId) {
        for (InputGroup inputGroup : InputSDKProvider.sGameInputMap.inputGroups()) {
            for (InputAction inputAction : inputGroup.inputActions()) {
                if (inputAction.inputActionId().uniqueId() == inputActionId) {
                    return Optional.of(inputAction);
                }
            }
        }
        return Optional.empty();
    }

    private Optional<InputAction> getCurrentVersionInputActionFromPreviousVersion(
            long inputActionId, String previousVersion) {
        // TODO: game developer can add logic to this method to get the input action considering the
        //  diff between the current and previous InputMap (if any).
        return Optional.empty();
    }

    private String keyCodesToString(List<Integer> keyCodes) {
        StringBuilder builder = new StringBuilder();
        for (Integer keyCode : keyCodes) {
            if (!builder.toString().isEmpty()) {
                builder.append(" + ");
            }
            builder.append(keyCode);
        }
        return String.format("(%s)", builder);
    }
}
